builtin/getopts.c

```diff
@@ -0,0 +1,88 @@
+#define _POSIX_C_SOURCE 200809L
+#include <errno.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+#include <mrsh/buffer.h>
+#include <mrsh/shell.h>
+#include <limits.h>
+#include "builtin.h"
+
+static const char read_usage[] = "usage: getopts optstring name [arg...]\n";
+
+int builtin_getopts(struct mrsh_state *state, int argc, char *argv[]) {
+	if (argc < 3) {
+		fprintf(stderr, read_usage);
+		return EXIT_FAILURE;
+	}
+	
+	int optc;
+	char **optv;
+	if (argc > 3) {
+		optc = argc - 2;
+		optv = &argv[2];
+	} else {
+		optc = state->argc;
+		optv = state->argv;
+	}
+	
+	const char *optind_str = mrsh_env_get(state, "OPTIND", NULL);
+	if (optind_str == NULL) {
+		fprintf(stderr, "getopts: OPTIND is not defined\n");
+		return EXIT_FAILURE;
+	}
+	char *endptr;
+	long optind_int = strtol(optind_str, &endptr, 10);
+	if (endptr[0] != '\0' || optind_int <= 0 || optind_int > INT_MAX) {
+		fprintf(stderr, "getopts: OPTIND is not a positive integer\n");
+		return EXIT_FAILURE;
+	}
+	optind = (int)optind_int;
+	
+	optopt = 0;
+	int opt = getopt(optc, optv, argv[1]);
+	
+	char optind_fmt[16];
+	snprintf(optind_fmt, sizeof(optind_fmt), "%d", optind);
+	mrsh_env_set(state, "OPTIND", optind_fmt, MRSH_VAR_ATTRIB_NONE);
+	
+	if (optopt != 0) {
+		if (opt == ':') {
+			char value[] = {(char)optopt, '\0'};
+			mrsh_env_set(state, "OPTARG", value, MRSH_VAR_ATTRIB_NONE);
+		} else if (argv[1][0] != ':') {
```

Comment:
Is it possible to add a variable `optstring = argv[1]`? This would make it easier to follow the logic here.